### PR TITLE
[POC][PYTHON][SS] Arbitrary stateful processing in Structured Streaming with Python

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR adds the Python version of `Dataset.groupByKey(...).flatMapGroupsWithState(...)` that is `DataFrame.groupby(...).applyInPandasWithState(...)` in PySpark.

TBD

Note that documentation will be done in a separate PR given the size of the PR.

### Why are the changes needed?

TBD

### Does this PR introduce _any_ user-facing change?

Yes, this PR adds a new API `DataFrame.groupby(...).applyInPandasWithState(...)` in PySpark.

```python
import typing

import pandas as pd

from pyspark.sql.types import StructType, StructField, LongType, StringType
from pyspark.sql.streaming.state import GroupStateTimeout, GroupStateImpl

output_type = StructType([
    StructField("key", LongType()),
    StructField("countAsString", StringType())])
state_type = StructType([StructField("count", LongType())])

# Type hints are optional in `func`.
def func(key: typing.Tuple, pdf: pd.DataFrame, state: GroupStateImpl) -> pd.DataFrame:
    count = state.getOption
    if count is None:
        count = 0
    else:
        count = count[0]
    count += len(pdf)
    state.update((count,))
    return pd.DataFrame({'key': [key[0]], 'countAsString': [str(count)]})


df = spark.readStream.format("rate").option("rowsPerSecond", 10).load().selectExpr("value % 3 as v")
df.groupBy(df["v"]).applyInPandasWithState(
    func, output_type, state_type, "Update", GroupStateTimeout.NoTimeout
).writeStream.format("console").queryName("test").start()
```

### How was this patch tested?

Manually tested, unittests and e2e tests were added in both Python and Scala sides.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/FlatMapGroupsInPandasWithStateExec.scala [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        Array(argOffsets),
+        StructType.fromAttributes(dedupAttributes),
+        sessionLocalTimeZone,
+        pythonRunnerConf,
+        groupedState,
+        stateDeserializer,
+        stateType)
+
+      val inputIter = if (hasTimedOut) {
+        lazy val unsafeProj = UnsafeProjection.create(
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/FlatMapGroupsInPandasWithStateExec.scala:125]:
```suggestion
        val unsafeProj = UnsafeProjection.create(
```

**Comment by HyukjinKwon** [comment]
cc @viirya FYI who I think is the best one who can review. I will attach some design docs, etc soon around next week likely.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowPythonRunnerWithState.scala [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.spark.sql.api.python.PythonSQLUtils
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+import org.apache.spark.sql.execution.streaming.GroupStateImpl
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * [[ArrowPythonRunner]] with [[org.apache.spark.sql.streaming.GroupState]].
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowPythonRunnerWithState.scala:37]: Explain the (de)serialization format.

**Comment by HeartSaVioR** [comment]
We can close this now.

**Pull request closed without merging**